### PR TITLE
Remove empty rows and fix repeating events forms specifications

### LINKED ISSUE #136: [BUG] `read_redcap()` returns rows for all instruments in a repeating event regardless of whether there is data
# Expected Behavior
`read_redcap()` only returns records for an event when some data has been entered for the instrument associated with event.

# Current Behavior
This is the current behavior for nonrepeating events. However, for repeating events `read_redcap()` returns rows for each instrument-`​redcap_event`-`​redcap_event_instance` combination even if no data has been filled in.

## How to Reproduce the Bug:
This is the current setup for the first records in our deep dive vignette redcap:

Screenshot 2023-03-03 at 4 39 05 PM

There are two repeat instances of `unscheduled` but the `chemistry` instrument has not been filled out for either.

```r
devtools::load_all()
library(tidyverse)

out <- read_redcap(Sys.getenv("REDCAP_URI"), Sys.getenv("REDCAPTIDIER_DEEP_DIVE_VIGNETTE_API"))

# chemistry has records for unscheduled but not for nonrepeating events that it is associated with
out |>
  extract_tibble("chemistry") |>
  distinct(redcap_event) |>
  pull()

#> [1] "screening__enrollm" "preinfusion"        "infusion"           "1_day_followup"     "unscheduled"

out |>
  extract_tibble("chemistry") |>
  filter(redcap_event == "unscheduled")

#> # A tibble: 2 × 16
#> subject_id  redca…¹ redca…² chem_per…³ chem_…⁴ chem_…⁵ chem_…⁶ chem_…⁷ chem_…⁸ chem_…⁹ chem_…˟ chem_…˟ chem_…˟
#> <chr>       <chr>     <dbl> <date>       <dbl>   <dbl>   <dbl>   <dbl>   <dbl>   <dbl>   <dbl>   <dbl>   <dbl>
#>   1 001-123-001 unsche…       1 NA              NA      NA      NA      NA      NA      NA      NA      NA      NA
#>   2 001-123-001 unsche…       2 NA              NA      NA      NA      NA      NA      NA      NA      NA      NA
```

### Checklist
Before submitting this issue, please check and verify below that the submission meets the below criteria:

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [ ] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### LINKED ISSUE #137: [BUG] `forms` specification in read_redcap drops REDCap identifier columns
# Expected Behavior
Specifying a `forms` argument returns all applicable fields to the related forms as well as necessary columns for composite key identification.

# Current Behavior
I noticed when testing for #136, that if I specified `read_redcap(redcap_uri = redcap_uri, token = token, forms = c("physical_exam", "chemistry"))`, that `redcap_form_instance` and `redcap_event_instance` do not get properly created. With no specification, expected columns are created as expected.

It looks like the issue is here in `read_redcap`:

https://github.com/CHOP-CGTInformatics/REDCapTidieR/blob/31a7774e58eb8157a732f9c13893dd5afda6b8e3/R/read_redcap.R#L194-L197

Before the inclusion of repeating events support, we were identifying if the `redcap_repeat_instrument`/`redcap_repeat_instance` columns could be dropped after `forms` specification in the API based on all `NA` in the `redcap_repeat_instrument` column. Now, all `NA` is possible since repeat events are `NA` in this column, but have value in `redcap_repeat_instance`.

## How to Reproduce the Bug:

Using the diving deeper vignette API key, we can see below how specifying `forms` leads to removal of the necessary REDCap columns required for composite key identification. Here we would expect `redcap_event_instance` to be present in the outputs.

```r
library(devtools)
document()
load_all()

redcap_uri <- Sys.getenv("REDCAP_URI")
token <- Sys.getenv("REDCAPTIDIER_DEEP_DIVE_VIGNETTE_API")

out <- read_redcap(redcap_uri = redcap_uri, token = token, forms = c("physical_exam", "chemistry"))

> names(out$redcap_data[[1]])
 [1] "subject_id"                    "redcap_event"                  "physical_capture"             
 [4] "physical_capture_infusion_sql" "physical_capture_visit"        "physical_performed"           
 [7] "physical_performed_date"       "physical_performance_scale"    "physical_performance_score"   
[10] "physical_findings"             "physical_findings_specify"     "form_status_complete"         
> names(out$redcap_data[[2]])
 [1] "subject_id"           "redcap_event"         "chem_performed_date"  "chem_sodium_mmoll"   
 [5] "chem_potassium_mmoll" "chem_bun_mgdl"        "chem_creatinine_mgdl" "chem_glucose_mgdl"   
 [9] "chem_calcium_mgdl"    "chem_protein_gdl"     "chem_albumin_gdl"     "chem_phosphatase_ul" 
[13] "chem_alt_ul"          "chem_ast_ul"          "form_status_complete"
```

Without specifying a forms argument, the same names are shown below:

```r
out <- read_redcap(redcap_uri = redcap_uri, token = token)

> out %>%
+   extract_tibble("physical_exam") %>% 
+   names()
 [1] "subject_id"                    "redcap_event"                 
 [3] "redcap_event_instance"         "physical_capture"             
 [5] "physical_capture_infusion_sql" "physical_capture_visit"       
 [7] "physical_performed"            "physical_performed_date"      
 [9] "physical_performance_scale"    "physical_performance_score"   
[11] "physical_findings"             "physical_findings_specify"    
[13] "form_status_complete"         
> out %>%
+   extract_tibble("chemistry") %>%
+   names()
 [1] "subject_id"            "redcap_event"          "redcap_event_instance"
 [4] "chem_performed_date"   "chem_sodium_mmoll"     "chem_potassium_mmoll" 
 [7] "chem_bun_mgdl"         "chem_creatinine_mgdl"  "chem_glucose_mgdl"    
[10] "chem_calcium_mgdl"     "chem_protein_gdl"      "chem_albumin_gdl"     
[13] "chem_phosphatase_ul"   "chem_alt_ul"           "chem_ast_ul"          
[16] "form_status_complete" 
```

### Checklist
Before submitting this issue, please check and verify below that the submission meets the below criteria:

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [x] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### PR DESCRIPTION
**Pull request by rsh52**
# Description
This PR addresses two bugs.

The first bug is a simple fix related to the `forms` argument. Prior to the merge of #126, we had no need to be concerned with `redcap_repeat_instance` as a column when `redcap_repeat_instrument` was all `NA`. With the inclusion of repeating events, we now know that `redcap_repeat_instrument` can be all `NA`, but `redcap_repeat_instance` can still be an indicator for data. I didn't make a test for this because of how simple it is, and making a test for it would require mocks edits. But happy to pursue this if we find it necessary.

The second bug relates to how our API processes downstream tibbles. For longitudinal databases, the REDCap API produces rows based on the composite key of `record_id`, `redcap_event_name`, `redcap_repeat_instrument`, and `redcap_repeat_instance`. In the sparse block matrix each row will have some amount of data filled out to validate the row's inclusion in the export since the block matrix contains all columns in the database. However, when we separate the matrix into downstream tibbles, some of those rows may no longer have data. Refer to the examples in #136 for a good visual of how this is portrayed with instances under an event and instances under a repeating event.

To solve this, `filter(if_any())` helps remove empty rows for identified columns of interest. This only affects longitudinal databases.

# Proposed Changes 
List changes below in bullet format:

- Add `redcap_repeat_instance` as a qualifier for `forms` filtering to `read_redcap`
- Add new `remove_empty_rows` utility function to longitudinal `distill_*` functions
- Add associated tests

### Issue Addressed
**Closes #136**
**Closes #137**

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [NA] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [NA] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR
- [x] Pre-release package version incremented using `usethis::use_dev_version()`

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [x] I checked that new files follow naming conventions and are in the right place
- [x] I checked that documentation is complete, clear, and without typos
- [x] I added/edited comments to explain "why" not "how"
- [x] I checked that all new variable and function names follow naming conventions
- [x] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [x] I checked that new tests address important edge cases

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -689,3 +689,41 @@ try_redcapr <- function(expr, call = caller_env()) {
...
+#' pre-processed within a `distill_*` function.
+#' @param my_record_id The record ID defined in the project.
+#'
+#' @importFrom dplyr filter if_any
+#'
+#' @keywords internal
+
+remove_empty_rows <- function(data, my_record_id) {
+  # Define standard columns that do not impact anlaysis
+  standard_cols <- c(my_record_id,
```
Comment (skadauke) [inline R/utils.R:716]: Nit: could rename `standard_cols` to `nondata_cols`.

File: R/utils.R [hunk 2]
```diff
@@ -689,3 +689,41 @@ try_redcapr <- function(expr, call = caller_env()) {
...
+                     "redcap_event",
+                     "redcap_arm",
+                     "redcap_form_instance",
+                     "redcap_event_instance",
+                     "form_status_complete")
```
Comment (skadauke) [inline R/utils.R:721]: I think we need to include survey columns here as well: `redcap_survey_timestamp` and `redcap_survey_identifier`

File: R/utils.R [hunk 3]
```diff
@@ -689,3 +689,41 @@ try_redcapr <- function(expr, call = caller_env()) {
 db_has_repeat_forms <- function(db_data) {
   "redcap_repeat_instance" %in% names(db_data)
 }
+
+#' @title
+#' Remove rows with empty data
+#'
+#' @description
+#' Check for empty rows in columns that aren't necessary for composite key
```
Comment (skadauke) [inline R/utils.R:697]: I think the idea is that we want to remove rows that are empty in all [data columns](https://chop-cgtinformatics.github.io/REDCapTidieR/articles/glossary.html#data-column). These include columns necessary for composite key spec (i.e. [identifier columns](https://chop-cgtinformatics.github.io/REDCapTidieR/articles/glossary.html#identifier-column)) but also the survey and and form completion columns.

File: R/utils.R [hunk 4]
```diff
@@ -689,3 +689,41 @@ try_redcapr <- function(expr, call = caller_env()) {
...
+  # Define standard columns that do not impact anlaysis
+  standard_cols <- c(my_record_id,
+                     "redcap_event",
+                     "redcap_arm",
+                     "redcap_form_instance",
+                     "redcap_event_instance",
+                     "form_status_complete")
+
+  # Subset columns that do impact analysis
+  check_empty_cols <- names(data)[!names(data) %in% standard_cols]
```
Comment (skadauke) [inline R/utils.R:724]: Nit: for clarity, you could call this `data_cols`

**Review by skadauke — CHANGES_REQUESTED** [review]

File: R/clean_redcap_long.R [hunk 5]
```diff
@@ -19,7 +19,7 @@
 #' \code{form_name} and \code{structure} column details.
 #'
 #' @importFrom checkmate assert_data_frame
-#' @importFrom dplyr filter pull
+#' @importFrom dplyr filter pull if_any
```
[the remark this answers is not in the document]
Comment (rsh52) [reply R/clean_redcap_long.R:22]: You're right, originally had the utility function logic in these higher level ones and forgot to take it out. Good catch 👍

File: R/clean_redcap_long.R [hunk 6]
```diff
@@ -194,7 +194,8 @@ distill_nonrepeat_table_long <- function(form_name,
     rename("redcap_survey_timestamp" = any_of(paste0(my_form, "_timestamp"))) %>%
     relocate(any_of("redcap_survey_timestamp"), .after = everything()) %>%
     rename("form_status_complete" = paste0(my_form, "_complete")) %>%
-    relocate("form_status_complete", .after = everything())
+    relocate("form_status_complete", .after = everything()) %>%
+    remove_empty_rows(my_record_id)
```
[the remark this answers is not in the document]
Comment (rsh52) [reply R/clean_redcap_long.R:198]: I think either way works, I just like using the record ID explicitly in case for some reason the positioning ever wasn't the same.

File: R/utils.R (hunk 3 above)
Comment (rsh52) [reply R/utils.R:697]: Sounds good, changing the first sentence to `"Remove rows that are empty in all associated data columns (those derived from fields in REDCap)."`

**Review by skadauke — APPROVED** [review]
LGTM!
